paddle/fluid/compiler/piano/shape.h

```diff
@@ -0,0 +1,149 @@
+  // stores.
+  //
+  // Methods for accessing the element type.
+  note::ElementType element_type() const { return element_type_; }
+  void set_element_type(note::ElementType value) { element_type_ = value; }
+
+  // Methods for accessing the dimensions array.
+  std::vector<int64_t>* mutable_dimensions() { return &dimensions_; }
+  const std::vector<int64_t>& dimensions() const { return dimensions_; }
+
+  // Methods for accessing the tuple subshapes. This field only non-empty for
+  // tuple shapes.
+  std::vector<Shape>* mutable_tuple_shapes() { return &tuple_shapes_; }
+  const std::vector<Shape>& tuple_shapes() const { return tuple_shapes_; }
+
+  // Methods for accessing the layout field.
+  Layout* mutable_layout() { return &layout_; }
+  const Layout& layout() const { return layout_; }
+  bool has_layout() const { return layout_.Valid(); }
+
+ private:
+  // The element type of this shape (tuple, array, etc).
+  note::ElementType element_type_ = note::INVALID_ELEMENT_TYPE;
+
+  // The array bounds of the dimensions.
+  std::vector<int64_t> dimensions_;
+
+  // The tuple element subshapes. This is non-empty only for tuple shapes.
+  std::vector<Shape> tuple_shapes_;
+
+  // The layout of the shape. Only relevant for arrays.
+  Layout layout_;
+};
+
+// Shape of the parameters and output of a note funtion, as its signature.
+class Signature {
+ public:
+  Signature() = default;
+
+  // Creates a Signature from a SignatureProto protobuf.
+  explicit Signature(const note::SignatureProto& proto);
+
+  Signature(std::vector<Shape>&& parameters,
+            std::vector<std::string>&& parameter_names, const Shape& result)
+      : parameters_(parameters),
+        parameter_names_(parameter_names),
+        result_(result) {}
+
+  note::SignatureProto ToProto() const;
+
+  std::string ToString() const;
+
+  // The following methods for accessing the data member of a Signature object
+  // stores.
+  //
+  // Methods for accessing and manipulating the Shape of the parameters.
+  std::vector<Shape>* mutable_parameters() { return &parameters_; }
+  const std::vector<Shape>& parameters() const { return parameters_; }
+
+  // Methods for accessing and manipulating the Shape of the result.
+  Shape* mutable_result() { return &result_; }
+  const Shape& result() const { return result_; }
+
+  // Methods for accessing and manipulating the names of the parameters.
+  std::vector<std::string>* mutable_parameter_names() {
+    return &parameter_names_;
+  }
+  const std::vector<std::string>& parameter_names() const {
+    return parameter_names_;
+  }
+
+ private:
+  // The shapes of the parameters of the function represented by this object.
+  std::vector<Shape> parameters_;
+
+  // The names of the parameters of the function represented by this object.
+  std::vector<std::string> parameter_names_;
+
+  // The shape of the result of the function represented by this object.
+  Shape result_;
```

Comment:
Same as the problem I mentioned in `proto`, should we support multiple returns?

Reply:
Please refer to the answer on last comment